pad_server/src/controller.c

```diff
@@ -102,71 +108,83 @@ static ssize_t controller_recv(controller_t *controller, void *buf, size_t n) {
  * TODO: docs
  */
 void *controller_run(void *arg) {
-
     controller_args_t *args = (controller_args_t *)(arg);
     controller_t controller;
     int err;
 
-    /* Connect to the controller */
-    err = controller_init(&controller, args->port);
-    if (err) {
-        fprintf(stderr, "Could not initialize controller with error: %s\n", strerror(err));
-        thread_return(err);
-    }
     pthread_cleanup_push(controller_cleanup, &controller);
```

Comment:
Looks much better! There is just one error case to consider here left.

Right after `pthread_cleanup_push`, if the thread terminates for any reason, the disconnect function will still be
called on an uninitialized controller struct (from line 112). Here, the values of `sock` and `client` can be garbage
stack values which aren't guaranteed to be `< 0`. So the disconnect function will still attempt to close a socket which doesn't exist. It should be able to handle this, but ideally you could set `controller.sock` and `controller.client` to `-1` right after line 112 to be safe.

Reply:
Ok yes will do that